socks5.py

```diff
@@ -232,28 +273,64 @@ def tcp_loop(self, sock1, sock2):
                 if not data:
                     break
                 sock2.sendall(data)
+                with traffic_lock:
+                    outbound_traffic += len(data)
 
             if sock2 in r:
                 data = sock2.recv(4096)
                 if not data:
                     break
                 sock1.sendall(data)
+                with traffic_lock:
+                    inbound_traffic += len(data)
 
     def handle_connect(self, address, port):
         log_tag = '%s:%s -> %s:%s' % (self.client_address + (address, port))
-        try:
-            remote = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            if CONNECT_HOST:
-                remote.bind((CONNECT_HOST, 0))
-            remote.connect((address, port))
-            logging.info('%s: connected', log_tag)
-        except Exception as err:
-            logging.error('%s: connect error %s', log_tag, err)
-            self.send_reply(self.STATUS_ECONNREFUSED)
-            self.server.close_request(self.request)
-            return
 
-        self.send_reply(self.STATUS_SUCCEEDED)
+        if isinstance(address, dict):
+            ipv4 = address.get('ipv4')
+            ipv6 = address.get('ipv6')
+        else:
+            ipv4 = address
+            ipv6 = None
+
+        if ipv6 is not None or ipv4 is not None:
+            with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
+                if ipv6 is not None:
+                    ipv6_future = executor.submit(self.connect_to_address, ipv6, port, socket.AF_INET6, log_tag, timeout=5)
+                if ipv4 is not None:
+                    # Start IPv4 connection after 300ms delay
+                    ipv4_future = executor.submit(self.delayed_connect, ipv4, port, socket.AF_INET, 0.3, log_tag)
```

Comment:
This seems like it will affect the latency of IPv4 connections, which might be an unacceptable tradeoff (a very large number of people, myself included, are still in IPv4-only environments).

Reply:
Yeah. I don't like this as it is currently, and I'm working on improving this.

Reply:
As far as I can tell, this is a major refactor to get this fully the way it should be, as it will require async sockets since concurrent.futures doesn't allow for fixing this properly.

I use T-Mobile which is dual stack, so this is overall an improvement for me, but as you said, if you are IPv4-only it will be an overall downgrade for you.

While I do intend to take a look at cleaning this up properly, because of it's complexity, I may not be able to get this fixed quickly, so you'd be totally in the right to hold off on merging this in until then.

Thanks for all of your comments so far.

Reply:
@nneonneo can you please evaluate this as it is now? Assuming you don't find any other problems with it, the only remaining problem with the IPv6 handling I am aware of is that the Happy Eyeballs algorithm will handle all situations well except for cases where an AAAA record is set, but IPv6 goes to a black hole. A proper fix to this is a complex refactor of the whole code which I would like to avoid doing. I can add in a test to see if it can connect to www.google.com over ipv6 when it first launches, and if that fails, assume IPv6 does not work from then forward, which would avoid having a delay on an ongoing basis. [I'm not sure if you are comfortable with that as a fix or not]

Otherwise, one other issue I do know is still there, I haven't added counters into the UDP support. I'll get to that tomorrow.

Assuming we agree on those two things as the final corrections, after that I think it'd be good to merge.

Reply:
I added both the google test and the udp counters.

Just one important thing to note: I don’t have access to an IPv4-only cellular network, so I haven’t tested that use case at all… but I did test it on my laptop on an IPv4-only WiFi connection.